subsys/lorawan/services/frag_decoder_lowmem.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2024 A Labs GmbH
+ * Copyright (c) 2024 tado GmbH
+ * Copyright (c) 2022 Jiapeng Li
+ *
+ * Based on: https://github.com/JiapengLi/LoRaWANFragmentedDataBlockTransportAlgorithm
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef FRAG_DEC_H_
+#define FRAG_DEC_H_
+
+#include <stdint.h>
+#include <stddef.h>
+
+#define FRAG_MAX_NB                                                                                \
+	(CONFIG_LORAWAN_FRAG_TRANSPORT_IMAGE_SIZE / CONFIG_LORAWAN_FRAG_TRANSPORT_MIN_FRAG_SIZE +  \
+	 1U)
+#define FRAG_MAX_SIZE  (CONFIG_LORAWAN_FRAG_TRANSPORT_MAX_FRAG_SIZE)
+#define FRAG_TOLERANCE (FRAG_MAX_NB * CONFIG_LORAWAN_FRAG_TRANSPORT_MAX_REDUNDANCY / 100U)
+
+#define FRAG_DEC_ONGOING                 (-1)
+#define FRAG_DEC_ERR_INVALID_FRAME       (-2)
+#define FRAG_DEC_ERR_TOO_MANY_FRAME_LOST (-3)
+#define FRAG_DEC_ERR                     (-4)
+
+typedef struct {
+	/** number of fragments */
+	uint16_t nb_frag;
+	uint8_t frag_size;
+} frag_dec_cfg_t;
+
+typedef enum {
+	/* wait uncoded fragmentations */
+	FRAG_DEC_STA_UNCODED,
+	/* wait coded fragmentations, uncoded frags are processed as coded ones */
+	FRAG_DEC_STA_CODED,
+	FRAG_DEC_STA_DONE,
+} frag_dec_sta_t;
+
+typedef struct {
+	frag_dec_cfg_t cfg;
+	frag_dec_sta_t status;
+
+	uint16_t lost_frame_count;
+	uint16_t filled_lost_frm_count;
+} frag_dec_t;
```

Comment:
We typically avoid using typedefs for `struct`s and `enum`s in Zephyr (see [here](https://www.kernel.org/doc/html/v4.14/process/coding-style.html#typedefs) for some good reasons). Also I think we could avoid the additional struct for config and just put the two members into the decoder struct:

```suggestion
enum frag_decoder_status {
	/* wait for uncoded fragments */
	FRAG_DEC_STATUS_UNCODED,
	/* wait for coded fragments */
	FRAG_DEC_STATUS_CODED,
	FRAG_DEC_STATUS_DONE,
};

struct frag_decoder {
	uint16_t nb_frag;
	uint8_t frag_size;
	enum frag_decoder_status status;
	uint16_t lost_frame_count;
	uint16_t filled_lost_frm_count;
};
```